### PR TITLE
Cleanup of throws and asserts of goto_convert.cpp

### PR DESCRIPTION
**Pull request by NlightNFotis**
In the same spirit as #2904 I am taking over development of #2814 because the original repository is now defunct, and I can't go further with that PR. This PR will host further development of that line of work.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [n/a] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by tautschnig — CHANGES_REQUESTED** [review]
At the current state: please see https://github.com/diffblue/cbmc/pull/2814 for a number of blocking request.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: a6c16991dc331f1cba2c9633753a11f93ac46180).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85897734
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 3e57a9c8eb4149dd2190f87b115ec68f07eac0ea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85899011

**Comment by NlightNFotis** [comment]
@tautschnig This should be now in a state where it's ready for a review. Just a note:

I have tried my best to make the commits as close as possible to the contribution guidelines. It's a little bit hairy, because 3 people have worked on this thus far, and one of them is no longer with us, but we have to be careful not to lose commit information and attribution information, hence why the commits are the way they are and I can not squash them better.

File: regression/ansi-c/goto_convert/switch_range_bounds/test.desc [hunk 1]
```diff
@@ -0,0 +1,8 @@
+CORE
```
Comment (tautschnig) [inline regression/ansi-c/goto_convert/switch_range_bounds/test.desc]: Isn't this one GCC-only?

File: regression/ansi-c/goto_convert/invalid_goto_label/test.desc [hunk 2]
```diff
@@ -0,0 +1,8 @@
+CORE gcc-only
```
Comment (tautschnig) [inline regression/ansi-c/goto_convert/invalid_goto_label/test.desc]: Why is this gcc-only?

File: regression/ansi-c/goto_convert/switch_range_empty/test.desc [hunk 3]
```diff
@@ -0,0 +1,8 @@
+CORE
+main.c
+
+^EXIT=134$
```
Comment (tautschnig) [inline regression/ansi-c/goto_convert/switch_range_empty/test.desc]: Huh? That's not what should be happening...

File: src/goto-programs/goto_convert.cpp [hunk 4]
```diff
@@ -602,14 +573,7 @@ void goto_convertt::convert_expression(
...
 {
-  if(code.operands().size()!=1)
-  {
-    error().source_location=code.find_source_location();
-    error() << "expression statement takes one operand" << eom;
-    throw 0;
-  }
-
-  exprt expr=code.op0();
+  exprt expr = code.expression();
```
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp:576]: Just curious: why can't this be a const reference?

File: src/goto-programs/goto_convert.cpp [hunk 5]
```diff
@@ -1924,14 +1856,12 @@ irep_idt goto_convertt::get_string_constant(const exprt &expr)
...
 
-  if(get_string_constant(expr, result))
-  {
-    error().source_location=expr.find_source_location();
-    error() << "expected string constant, but got: "
-      << expr.pretty() << eom;
-
-    throw 0;
-  }
+  bool res = get_string_constant(expr, result);
```
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp]: `const`

File: regression/ansi-c/goto_convert/switch_range_bounds/test.desc (hunk 1 above)
Comment (NlightNFotis) [reply regression/ansi-c/goto_convert/switch_range_bounds/test.desc]: No, both `clang` and `gcc` report the same error.

File: regression/ansi-c/goto_convert/invalid_goto_label/test.desc (hunk 2 above)
Comment (NlightNFotis) [reply regression/ansi-c/goto_convert/invalid_goto_label/test.desc]: You're right, this is a mistake, both `clang` and `gcc` report the same error.

File: regression/ansi-c/goto_convert/switch_range_bounds/test.desc (hunk 1 above)
Comment (tautschnig) [reply regression/ansi-c/goto_convert/switch_range_bounds/test.desc]: Ok, but Visual Studio might not? But either way this one should fail in the front-end, so it's probably ok.

File: regression/ansi-c/goto_convert/switch_range_empty/test.desc (hunk 3 above)
Comment (NlightNFotis) [reply regression/ansi-c/goto_convert/switch_range_empty/test.desc]: What should be happening? I see that `gcc` and `clang` produce warnings that the range is empty, but are happy to compile nonetheless. Should I remove the invariant?

File: src/goto-programs/goto_convert.cpp (hunk 4 above)
Comment (NlightNFotis) [reply src/goto-programs/goto_convert.cpp:576]: To my understanding it gets modified in the alternative branch of the if statement, by `clean_expr` and the rest of the code.

File: regression/ansi-c/goto_convert/switch_range_empty/test.desc (hunk 3 above)
Comment (tautschnig) [reply regression/ansi-c/goto_convert/switch_range_empty/test.desc]: We should be able to support whatever GCC and Clang do. 134 (a failing invariant) is not an exit code we should ever be producing for a syntactically valid C program.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9a6e16fadcec4108da63658f54fc58c3b0f41dc8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86143603

**Comment by NlightNFotis** [comment]
@tautschnig I decided on an alternative approach to the bugged test (the one we had a divergent behaviour from `gcc` and `clang`).

Before the refactoring work, that turned a line in `goto_convertt::convert_switch` in line 1252 from an `assert` to a `DATA_INVARIANT`. It seems that both of these behaviours are wrong to begin with, and I consider this a bug. For this reason I added a `KNOWNBUG` test, that I will mark as `CORE` in another PR, that I will submit on top of this PR fixing this, and in which I will only have the removal of the assertion and the activation of the test. So consider this one at a reviewable state.

**Comment by NlightNFotis** [comment]
I pushed #3055 that fixes this the issue suggested above. That PR also for now contains the changes introduced here, but once this one is merged, I will rebase that one appropriately to only mirror the change it introduces.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 5e7e55ad8902b8873f910d5b88e36bc56bc1aefc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86177330
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 552381869c6f82102a7233ce69b951e92d8b8cb6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86177840

**Review by tautschnig — APPROVED** [review]

**Pull request merged**